### ISSUE TITLE
sql: make `/` operator perform integer division

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/cockroach#41448]
Many ORM tests (e.g. Hibernate) expect the `/` operator to return an `int`, but CockroachDB currently returns a `decimal`. This was previously discussed in #3271 but we are revisiting the discussion now.

**Comment by otan** [comment]
So am I correct to assume this is a goahead because of our pivot to match postgres as much as possible?

**Comment by jordanlewis** [comment]
Yes.

**Comment by otan** [comment]
Updating this a little bit. The straightforward fix I had in mind (changing the `Div` to take in `Int`, `Int` and return `Int`) didn't work for cases like `select 1 / 5` (returns 0.2 instead of 0) due to the constant folding we do. Fixing or trying to alter that logic involves even more bugs / hairyness to colexec and tuple matching which I haven't really dug into. Do we want to take these kind of changes separately? Is there some larger discussion I'm missing?

**Comment by jordanlewis** [comment]
There's no larger discussion, as of yet. The investigation you're doing now is part of uncovering the unknowns here - I don't think anyone was fully sure what would happen if we started down this path. So it's definitely possible (likely?) that there won't be a quick fix here given what you've uncovered.

Perhaps the larger discussion that we'd want to start at some point soon is whether we should kind of redo our type system now that we've decided to flip the needle toward Postgres compatibility.

One practical question is whether our parser-level constant folding is even offering that much anymore - the optimizer also performs a constant folding pass. @RaduBerinde do you have an opinion on getting rid of the parser-level constant folding? My gut feel is that we could do it and it wouldn't make much of a difference to anybody.

**Comment by knz** [comment]
> you have an opinion on getting rid of the parser-level constant folding?

Just FWIW the folding is necessary to solve a number of type ambiguities.

> whether we should kind of redo our type system now that we've decided to flip the needle toward Postgres compatibility. 

:+1:

**Comment by otan** [comment]
> Just FWIW the folding is necessary to solve a number of type ambiguities.

can you give an example? is there a primer of how type ambiguities are resolved somewhere? thanks!

**Comment by knz** [comment]
Check the typing RFC it contains such examples.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mgartner** [comment]
This causes errors when trying to run `pgbench`. See #94238.

**Comment by rafiss** [comment]
I wonder if we can make this configurable with a session setting. I guess that would change the volatility of the `/` and `//` operators though, which could be bad.

**Comment by mgartner** [comment]
Good point. In this case we'd usually want to make `/` stable to prevent the optimizer from constant folding it during `PREPARE`, but I'm not sure that's necessary. We have to type check the expression differently, so any prepared statement in the query cache would have to be invalidated when the session setting changes anyway. 

But the fact that the behavior of `/` changes based on a session setting means that it can corrupt existing computed columns, partial indexes, and expression indexes, regardless of if we label it stable or leave it immutable.

So, rather than make it configurable with a session setting, I propose that we do the following in a new major version:

1. Create a new built-in that has the same type signature as the current form of `/`, `(int, int) -> decimal`.
2. Write a migration to convert `/` to this new builtin in any existing schema expressions.
3. Change the signature of `/` to `(int, int) -> int`.

The downside of this approach is that it can break queries, but I think that's acceptable in a major version upgrade.
